### PR TITLE
refactor: use `pkgs.writeShellApplication`

### PR DESCRIPTION
**Pull request by nekowinston**
Following #1, I had a look through other todo notes, and this would be my approach.

Tried it on my dots again. ~~Seems that only the formatting changed.~~ (`cfg.env` changed, see my comments)

writeShellApplication already runs shellcheck for us: https://github.com/NixOS/nixpkgs/blob/52544c4a0a84552db4013a8858ade52eaf7ff0ca/pkgs/build-support/trivial-builders/default.nix#L243-L357

### CODE REVIEW

File: modules/default.nix [hunk 1]
```diff
@@ -4,58 +4,28 @@
...
-      name = "neovim-env";
-      paths = [wrapper];
+    pkgs.writeShellApplication {
+      name = "nvim";
+      runtimeInputs = cfg.paths;
+      runtimeEnv =
+        {
+          NVIM_RPLUGIN_MANIFEST = "${config.neovim.build.rplugin}/rplugin.vim";
+        }
+        // (cfg.env or {});
```
Comment (nekowinston) [inline modules/default.nix]: Might want to flip this so NVIM_RPLUGIN_MANIFEST cannot be overriden?

Comment (nekowinston) [inline modules/default.nix]: The current code produces
```sh
export FOO="${FOO:-bar}"
```
while `runtimeEnv` produces
```sh
FOO='bar'
export FOO
```

**Comment by willruggiano** [comment]
So this one is actually on purpose. The reason is that `writeShellApplication` *prepends* PATH arguments (`export PATH="${lib.makeBinPath runtimeInputs}:$PATH"`) which means that paths can't be overridden (e.g. via direnv/devShell) if you include in them in `paths` :(

I had planned on submitting an upstream PR to allow *appending* rather than prepending to PATH, but guess I just forgot :/

**Comment by nekowinston** [comment]
Interesting, I didn't think of this use case before, but that makes a lot of sense.

Would going back to the previous way of setting `$PATH` work? https://github.com/willruggiano/neovim.nix/commit/e0926936f00582ea059de24a7733608ef3d40806

**Comment by willruggiano** [comment]
> Interesting, I didn't think of this use case before, but that makes a lot of sense.
> 
> Would going back to the previous way of setting `$PATH` work? [e092693](https://github.com/willruggiano/neovim.nix/commit/e0926936f00582ea059de24a7733608ef3d40806)

No, doing so is the equivalent of:
```
export PATH="${makeBinPath ...}:$PATH" # added by writeShellApplication
export PATH="$PATH:${makeBinPath ...}" # added by us
```

but when the second export hits, the damage has already been done by `writeShellApplication`.

**Comment by nekowinston** [comment]
The first line
```
export PATH="${makeBinPath ...}:$PATH" # added by writeShellApplication
```
doesn't get added for me when `runtimeInputs == []`, see https://github.com/NixOS/nixpkgs/blob/52544c4a0a84552db4013a8858ade52eaf7ff0ca/pkgs/build-support/trivial-builders/default.nix#L330-L333

When I'm building my neovim.drv with https://github.com/willruggiano/neovim.nix/commit/e0926936f00582ea059de24a7733608ef3d40806, I get:
```
#!/nix/store/0wzb184kckhmggq1nlvy5xk6rykp0jl1-bash-5.2-p21/bin/bash
set -o errexit
set -o nounset
set -o pipefail
NVIM_RPLUGIN_MANIFEST='/nix/store/6xyawv9vc956jxbmbb5km15r07z8bxjj-rplugin.vim/rplugin.vim'
export NVIM_RPLUGIN_MANIFEST

export PATH="$PATH:/nix/store/dycm8fa8rhxjyfra67s8mnqg4vmig3h3-fd-9.0.0/bin:..."
/nix/store/wfwz9ig0ipl21v1ngxhgndfgsix5h392-neovim-unwrapped-0.9.5/bin/nvim -u /nix/store/m9ji2sl0h44w3831xxpzkpzw6a7d4hk6-init.lua "$@"
```

And starting `PATH=FOO ./result/bin/nvim +'echo $PATH'` prepends `$PATH` for me: `FOO:/nix/store/dycm8fa8rhxjyfra67s8mnqg4vmig3h3-fd-9.0.0/bin:...`.

...sorry if this is coming across as trying to force this PR, just trying to figure out if something changed recently, that leads to different behavior for my build.

**Comment by willruggiano** [comment]
> The first line
> 
> ```
> 
> export PATH="${makeBinPath ...}:$PATH" # added by writeShellApplication
> 
> ```
> 
> doesn't get added for me when `runtimeInputs == []`, see https://github.com/NixOS/nixpkgs/blob/52544c4a0a84552db4013a8858ade52eaf7ff0ca/pkgs/build-support/trivial-builders/default.nix#L330-L333
> 
> 
> 
> When I'm building my neovim.drv with https://github.com/willruggiano/neovim.nix/commit/e0926936f00582ea059de24a7733608ef3d40806, I get:
> 
> ```
> 
> #!/nix/store/0wzb184kckhmggq1nlvy5xk6rykp0jl1-bash-5.2-p21/bin/bash
> 
> set -o errexit
> 
> set -o nounset
> 
> set -o pipefail
> 
> NVIM_RPLUGIN_MANIFEST='/nix/store/6xyawv9vc956jxbmbb5km15r07z8bxjj-rplugin.vim/rplugin.vim'
> 
> export NVIM_RPLUGIN_MANIFEST
> 
> 
> 
> export PATH="$PATH:/nix/store/dycm8fa8rhxjyfra67s8mnqg4vmig3h3-fd-9.0.0/bin:..."
> 
> /nix/store/wfwz9ig0ipl21v1ngxhgndfgsix5h392-neovim-unwrapped-0.9.5/bin/nvim -u /nix/store/m9ji2sl0h44w3831xxpzkpzw6a7d4hk6-init.lua "$@"
> 
> ```
> 
> 
> 
> And starting `PATH=FOO ./result/bin/nvim +'echo $PATH'` prepends `$PATH` for me: `FOO:/nix/store/dycm8fa8rhxjyfra67s8mnqg4vmig3h3-fd-9.0.0/bin:...`.
> 
> 
> 
> ...sorry if this is coming across as trying to force this PR, just trying to figure out if something changed recently, that leads to different behavior for my build.

Cool I will try it out when I get home!

**Pull request merged**